File: api-node/src/models/Repository.ts
import { Schema, model, Document, Types } from "mongoose";
import { RunMode } from "../types/enums";

export interface IRepository extends Document {
  orgId: Types.ObjectId;
  githubRepoId: number;
  name: string;
  fullName: string;
  description?: string;
  defaultBranch: string;
  language?: string;
  isPrivate: boolean;
  isEnabled: boolean;
  webhookId?: number;
  lastSyncedAt?: Date;
  config: {
    maxComments: number;
    minSeverity: "block" | "high" | "medium" | "low";
    shadowMode: boolean;
    runMode: RunMode;
    enableStatic: boolean;
    enableAi: boolean;
    aiMode: "balanced" | "strict" | "security" | "performance";
    aiModelOverride?: string;
    excludedPaths: string[];
    excludedFilePatterns: string[];
    maxFileSizeKb: number;
    maxPrFiles: number;
    maxPrLines: number;
    enabledRules: string[];
    disabledRules: string[];
    customRules: any[];
  };
  createdAt: Date;
  updatedAt: Date;
}

const repositorySchema = new Schema<IRepository>(
  {
    orgId: {
      type: Schema.Types.ObjectId,
      ref: "Organization",
      required: true,
      index: true,
    },
    githubRepoId: { type: Number, required: true, unique: true, index: true },
    name: { type: String, required: true },
    fullName: { type: String, required: true, index: true },
    description: { type: String },
    defaultBranch: { type: String, default: "main" },
    language: { type: String },
    isPrivate: { type: Boolean, default: false },
    isEnabled: { type: Boolean, default: false, index: true },
    webhookId: { type: Number },
    lastSyncedAt: { type: Date },
    config: {
      maxComments: { type: Number, default: 10 },
      minSeverity: {
        type: String,
        enum: ["block", "high", "medium", "low"],
        default: "low",
      },
      shadowMode: { type: Boolean, default: true },
      runMode: {
        type: String,
        enum: Object.values(RunMode),
        default: RunMode.SHADOW,
      },
      enableStatic: { type: Boolean, default: true },
      enableAi: { type: Boolean, default: true },
      aiMode: {
        type: String,
        enum: ["balanced", "strict", "security", "performance"],
        default: "balanced",
      },
      aiModelOverride: { type: String },
      excludedPaths: { type: [String], default: [] },
      excludedFilePatterns: {
        type: [String],
        default: [
          "*.lock",
          "*.min.js",
          "*.min.css",
          "package-lock.json",
          "yarn.lock",
        ],
      },
      maxFileSizeKb: { type: Number, default: 500 },
      maxPrFiles: { type: Number, default: 100 },
      maxPrLines: { type: Number, default: 5000 },
      enabledRules: { type: [String], default: [] },
      disabledRules: { type: [String], default: [] },
      customRules: { type: [Schema.Types.Mixed], default: [] },
    },
  },
  {
    timestamps: true,
  }
);

export const Repository = model<IRepository>("Repository", repositorySchema);

File: worker-python/src/models/Repository.py
from datetime import datetime
from typing import Optional, List
from beanie import Document, Link
from pydantic import Field, BaseModel
from .Organization import Organization

import pymongo

class RepoConfig(BaseModel):
    run_mode: str = Field("shadow", alias="runMode")
    max_pr_files: int = Field(100, alias="maxPrFiles")
    max_pr_lines: int = Field(5000, alias="maxPrLines")
    enable_ai: bool = Field(True, alias="enableAi")
    ai_mode: str = Field("balanced", alias="aiMode")
    ai_model_override: Optional[str] = Field(None, alias="aiModelOverride")
    excluded_paths: List[str] = Field(default_factory=list, alias="excludedPaths")
    excluded_file_patterns: List[str] = Field(default_factory=list, alias="excludedFilePatterns")
    enabled_rules: Optional[List[str]] = Field(None, alias="enabledRules")
    disabled_rules: Optional[List[str]] = Field(None, alias="disabledRules")
    max_comments: int = Field(10, alias="maxComments")
    min_severity: str = Field("low", alias="minSeverity")

class Repository(Document):
    org_id: Link[Organization] = Field(..., alias="orgId")
    github_repo_id: int = Field(..., alias="githubRepoId")
    name: str
    full_name: str = Field(..., alias="fullName")
    description: Optional[str] = None
    default_branch: str = Field(..., alias="defaultBranch")
    language: Optional[str] = None
    is_private: bool = Field(..., alias="isPrivate")
    is_enabled: bool = Field(True, alias="isEnabled")
    last_synced_at: Optional[datetime] = Field(None, alias="lastSyncedAt")
    config: RepoConfig = Field(default_factory=RepoConfig)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "repositories"
        indexes = [
            "orgId",
            pymongo.IndexModel([("githubRepoId", pymongo.ASCENDING)], unique=True),
            pymongo.IndexModel([("fullName", pymongo.ASCENDING)], unique=True),
        ]
